maze_dataset/tokenization/maze_tokenizer.py

```diff
@@ -388,3 +429,723 @@ def clear_cache(self):
+        """
+        # delete the properties only if they exist
+        for name, prop in self.__class__.__dict__.items():
+            if (
+                isinstance(prop, TokenizerElement) and recursive
+            ):  # Recursively clear cache for nested properties
+                prop.clear_cache()
+            if isinstance(prop, cached_property):
+                # if the property exists, delete it
+                try:
+                    delattr(self, name)
+                except AttributeError as e:
+                    pass
+
+
+class _TokenizerElementNamespace(abc.ABC):
+    """ABC for namespaces
+
+    # Properties
+    - key: The binding used in `MazeTokenizer2` for instances of the classes contained within that `_TokenizerElementNamespace`.
+    """
+
+    key: str = NotImplementedError
+
+
+class CoordTokenizers(_TokenizerElementNamespace):
+    key = "coord_tokenizer"
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class CoordTokenizer(TokenizerElement, abc.ABC):
+        @abc.abstractmethod
+        def to_tokens(self, coord: Coord | CoordTup) -> list[str]:
+            pass
+
+        # Define some (abstract) methods
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class UT(CoordTokenizer, abc.ABC):
+        def to_tokens(self, coord: Coord | CoordTup) -> list[str]:
+            return ["".join(["(", str(coord[0]), ",", str(coord[1]), ")"])]
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class CTT(CoordTokenizer):
+        """Coordinate tuple tokenizer
+
+        # Parameters
+        - `pre`: Whether all coords include an integral preceding delimiter token
+        - `intra`: Whether all coords include a delimiter token between coordinates
+        - `post`: Whether all coords include an integral following delimiter token
+        """
+
+        pre: bool = serializable_field(default=True)
+        intra: bool = serializable_field(default=True)
+        post: bool = serializable_field(default=True)
+        # Implement methods
+
+        def to_tokens(self, coord: Coord | CoordTup) -> list[str]:
+            return [
+                *unpackable_if_true_attribute([VOCAB.COORD_PRE], self, "pre"),
+                str(coord[0]),
+                *unpackable_if_true_attribute([VOCAB.COORD_INTRA], self, "intra"),
+                str(coord[1]),
+                *unpackable_if_true_attribute([VOCAB.COORD_POST], self, "post"),
+            ]
+
+
+class AdjListTokenizers(_TokenizerElementNamespace):
+    key = "adj_list_tokenizer"
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class AdjListTokenizer(TokenizerElement, abc.ABC):
+        """
+        Specifies how the adjacency list is tokenized.
+        """
+
+        @abc.abstractmethod
+        def to_tokens(self, conn_list: ConnectionList) -> list[str]:
+            pass
+
+        # Define some (abstract) methods
+
+    # TODO: figure out properties_to_serialize
```

Comment:
probably can remove this TODO, no properties

Reply:
Yep, delete it